File: python/2311.longest-binary-subsequence-less-than-or-equal-to-k.py
class Solution:
    def longestSubsequence(self, s: str, k: int) -> int:
        taken = 0
        zeros = s.count("0")
        n = len(s)
        cur_len = 0
        cur_val = 0

        for i in range(n - 1, -1, -1):
            if s[i] == "0":
                taken += 1
                zeros -= 1
            elif s[i] == "1":
                cur_val += 1 << cur_len

            if cur_val > k:
                return cur_len + zeros

            cur_len += 1

        return n

File: golang/2311.longest-binary-subsequence-less-than-or-equal-to-k.go
func longestSubsequence(s string, k int) int {
	n := len(s)
	curLen, curVal := 0, 0
	zeros := 0

	for i := range n {
		if s[i] == '0' {
			zeros++
		}
	}

	for i := n - 1; i >= 0; i-- {
		switch s[i] {
		case '0':
			zeros--
		case '1':
			if curLen <= 30 {
				curVal += 1 << curLen
			} else {
				return curLen + zeros
			}
		}

		if curVal > k {
			return curLen + zeros
		}

		curLen++
	}

	return n
}
